catalog/rest_test.go

```diff
@@ -0,0 +1,818 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package catalog_test
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/apache/iceberg-go"
+	"github.com/apache/iceberg-go/catalog"
+	"github.com/apache/iceberg-go/table"
+	"github.com/stretchr/testify/suite"
+)
+
+const (
+	TestCreds = "client:secret"
+	TestToken = "some_jwt_token"
+)
+
+var (
+	TestHeaders = http.Header{
+		"X-Client-Version": {"0.14.1"},
+		"User-Agent":       {"GoIceberg/" + iceberg.Version()},
+		"Authorization":    {"Bearer " + TestToken},
+	}
+	OAuthTestHeaders = http.Header{
+		"Content-Type": {"application/x-www-form-urlencoded"},
+	}
+)
+
+type RestCatalogSuite struct {
+	suite.Suite
+
+	srv *httptest.Server
+	mux *http.ServeMux
+
+	configVals url.Values
+}
+
+func (r *RestCatalogSuite) SetupTest() {
+	r.mux = http.NewServeMux()
+
+	r.mux.HandleFunc("/v1/config", func(w http.ResponseWriter, req *http.Request) {
+		r.Require().Equal(http.MethodGet, req.Method)
+		r.configVals = req.URL.Query()
+
+		json.NewEncoder(w).Encode(map[string]any{
+			"defaults":  map[string]any{},
+			"overrides": map[string]any{},
+		})
+	})
+
+	r.srv = httptest.NewServer(r.mux)
+}
+
+func (r *RestCatalogSuite) TearDownTest() {
+	r.srv.Close()
+	r.srv = nil
+	r.mux = nil
+	r.configVals = nil
+}
+
+func (r *RestCatalogSuite) TestToken200() {
+	r.mux.HandleFunc("/v1/oauth/tokens", func(w http.ResponseWriter, req *http.Request) {
+		r.Equal(http.MethodPost, req.Method)
+
+		r.Equal(req.Header.Get("Content-Type"), "application/x-www-form-urlencoded")
+
+		r.Require().NoError(req.ParseForm())
+		values := req.PostForm
+		r.Equal(values.Get("grant_type"), "client_credentials")
+		r.Equal(values.Get("client_id"), "client")
+		r.Equal(values.Get("client_secret"), "secret")
+		r.Equal(values.Get("scope"), "catalog")
+
+		w.WriteHeader(http.StatusOK)
+
+		json.NewEncoder(w).Encode(map[string]any{
+			"access_token":      TestToken,
+			"token_type":        "Bearer",
+			"expires_in":        86400,
+			"issued_token_type": "urn:ietf:params:oauth:token-type:access_token",
+		})
+	})
+
+	cat, err := catalog.NewRestCatalog("rest", r.srv.URL,
```

Comment:
I believe this needs to make sure that the correct authorization header is set, similar to https://github.com/apache/iceberg-python/blob/main/tests/catalog/test_rest.py#L137

Reply:
It's verified in the `TestListTablesPrefixed200` test, it uses the token check and then makes a request afterwards which confirms the authorization header is sent with the subsequent request.

Reply:
given that we're testing a different endpoint here, I would say it's better to also make sure that the correct Authorization header is set

Reply:
added test